### PR TITLE
Improve inspec shell by having it evaluate describe/control blocks.

### PR DESCRIPTION
**Pull request by ksubrama**
@arlimus @chris-rock @jbussdieker @alexpop 

This definitely has bugs in it but the is the general strategy I came up with last night after diving head-first into inspec and pry.

### CODE REVIEW

**Comment by ksubrama** [comment]
Note - the "-c" arm of the shell already works (I have tested it locally).  I'm trying to figure out how to make the shell work.

File: lib/bundles/inspec-supermarket/target.rb [hunk 1]
```diff
@@ -13,11 +13,11 @@ class Fetcher < Fetchers::Url
     priority 500
 
     def self.resolve(target, opts = {})
-      return nil unless URI(target).scheme == 'supermarket'
+      return nil unless target.is_a?(String)
```
Comment (chris-rock) [inline lib/bundles/inspec-supermarket/target.rb:16]: we still need to ensure that the scheme is `supermarket`

File: lib/bundles/inspec-compliance/target.rb [hunk 2]
```diff
@@ -15,6 +15,7 @@ class Fetcher < Fetchers::Url
     priority 500
 
     def self.resolve(target, _opts = {})
+      return nil unless target.is_a?(String)
```
Comment (chris-rock) [inline lib/bundles/inspec-compliance/target.rb:18]: 👍

File: lib/inspec/runner_rspec.rb [hunk 3]
```diff
@@ -87,17 +87,17 @@ def report
...
-    private
-
     # Empty the list of registered tests.
     #
     # @return [nil]
     def reset_tests
       @tests = RSpec::Core::World.new
       # resets "pending examples" in reporter
-      RSpec.configuration.reset
+      # RSpec.configuration.reset
```
Comment (chris-rock) [inline lib/inspec/runner_rspec.rb]: why is this not executed anymore and moved to `configure_output`?

File: lib/inspec/cli.rb [hunk 4]
```diff
@@ -185,9 +184,9 @@ def version
   private
 
   def run_command(opts)
-    opts[:test_collector] = Inspec::RunnerMock.new
     runner = Inspec::Runner.new(opts)
-    runner.create_context.load(opts[:command])
+    runner.add_target({'in_mem.rb' => opts[:command]}, opts)
```
Comment (chris-rock) [inline lib/inspec/cli.rb]: if we add a name, we should name it `in_memory.rb`

File: lib/inspec/cli.rb [hunk 5]
```diff
@@ -153,19 +153,18 @@ def detect
...
     diagnose
     o = opts.dup
     o[:logger] = Logger.new(STDOUT)
     o[:logger].level = get_log_level(o.log_level)
     if o[:command].nil?
       runner = Inspec::Runner.new(o)
-      return Inspec::Shell.new(runner).start
+      return Inspec::Shell.new(runner).start(o)
     else
-      res = run_command(o)
```
Comment (chris-rock) [inline lib/inspec/cli.rb:168]: I am not sure we should remove that. cc @arlimus

File: lib/inspec/shell.rb [hunk 6]
```diff
@@ -3,21 +3,28 @@
...
-    def start
-      # store context to run commands in this context
-      c = { content: 'binding.pry', ref: nil, line: nil }
-      @runner.add_content(c, [])
-      @runner.run
+    def start(opts)
+      # Create an in-memory empty profile so that we can add tests to it later.
+      # This context lasts for the duration of this "start" method call/pry
+      # session.
+      @ctx = @runner.add_target({'shell_context.rb' => ''}, opts)
```
Comment (chris-rock) [inline lib/inspec/shell.rb]: I am having some concerns with adding empty mock files. Investigating alternative solutions ...

**Comment by ksubrama** [comment]
Fix #240

File: lib/inspec/runner_rspec.rb (hunk 3 above)
Comment (ksubrama) [inline lib/inspec/runner_rspec.rb]: I restructured this.  Essentially - I wanted to reset_tests in the test_collector without resetting the output configuration and json formatter stuff.  But I found a more explicit way to do that.

File: lib/inspec/cli.rb [hunk 7]
```diff
@@ -163,11 +163,9 @@ def shell_func
     o[:logger].level = get_log_level(o.log_level)
     if o[:command].nil?
       runner = Inspec::Runner.new(o)
-      return Inspec::Shell.new(runner).start
+      return Inspec::Shell.new(runner, o).start
     else
-      res = run_command(o)
```
Comment (chris-rock) [inline lib/inspec/cli.rb:168]: we should reflect the `--format` option here and do the json dump if activated

**Comment by chris-rock** [comment]
This fixes #240

File: lib/inspec/cli.rb (hunk 7 above)
Comment (ksubrama) [inline lib/inspec/cli.rb:168]: run_command should already print the json output.  It's no longer using a mock runner - it's doing the real rspec thing which handles --format on its own.

**Comment by chris-rock** [comment]
Some test cases for that ticket. We should add an automatic test for the shell in another iteration.
```
# simple resources
file('/tmp').exist?
file('tmp/file').exist?

describe file('/tmp') do
  it { should exist}
end

describe file('tmp/file') do
  it { should_not exist}
end

control 'tmp' do
  title 'tmp directory exists'
  describe file('/tmp') do
    it { should exist}
  end
end

control 'tmp2' do
  title 'tmp/file does not exist'
  describe file('tmp/file') do
    it { should_not exist}
  end
end

res = describe file('tmp/file') do
  it { should_not exist}
end

shell with json output
b inspec shell -c "passwd.params" --format json
b inspec shell -c "file('/tmp')" --format json


# two controls with same name
control 'tmp2' do
  title 'tmp/file does not exist'
  describe file('tmp/file') do
    it { should_not exist}
  end
end
control 'tmp2' do
  title 'tmp/file does not exist'
  describe file('tmp/file') do
    it { should_not exist}
  end
end

# result via -c
inspec shell -c 'describe file("/Users/ksubramanian") do it { should exist } end'
```

**Pull request merged**